src/constraints.jl

```diff
@@ -577,8 +577,10 @@ representing the function and the `set` field contains the MOI set.
 See also the [documentation](@ref Constraints) on JuMP's representation of
 constraints for more background.
 """
-struct ScalarConstraint{F<:AbstractJuMPScalar,S<:MOI.AbstractScalarSet} <:
-       AbstractConstraint
+struct ScalarConstraint{
+    F<:Union{Number,AbstractJuMPScalar},
```

Comment:
Ok so this is needed for this test for instance:
https://github.com/jump-dev/JuMP.jl/blob/2d666cb01857701e7442cee844ba00ee83b4dca9/test/test_constraint.jl#L103
Before, we would just create an `AffExpr` but in `build_constraint`, we cannot build an `AffExpr` since we don't know what the variable type is since it depends on the value type and we don't have any variable nor model.
Actually, building an `AffExpr` is not so extension friendly either, e.g., the `JuMPExtension` also has a different variable type so I think it does make sense to allow `Number` here.